### PR TITLE
<va-file-input>: Update focus management for errors and interactive child elements

### PR DESCRIPTION
**Pull request by RyanMunsch**
## Chromatic

https://file-input-focus-and-sr--65a6e2ed2314f7b8f98609d8.chromatic.com

## Description

This pull request introduces several improvements and new features to the `va-file-input` component, focusing on accessibility, user experience, and developer utilities. The most significant changes include an overhaul to screen reader support, changes for which led to the addition of a comprehensive utility module for the component, enhanced test coverage (especially for accessibility features), and expanded Storybook stories to demonstrate new behaviors and requirements.

**Key changes:**

### Accessibility and UX Improvements

* Added a new utility module `utils.tsx` for `va-file-input` containing helper functions for accessibility (e.g., focusing elements, generating aria-labels, formatting file sizes, rendering error alerts/thumbnails, and more). This greatly improves the maintainability and consistency of accessibility features across the component.
* Expanded end-to-end tests to verify that the file input's `aria-label` is updated correctly when files are uploaded or deleted, and that error messages are properly rendered and accessible.
* Adjusted test timing and selectors to improve reliability and ensure correct focus and modal behaviors in edge cases. [[1]](diffhunk://#diff-d9221f0cd062e8c59fb880be651ebe48cb2182116c9e07fbc1f364794e8b28dcL219-R218) [[2]](diffhunk://#diff-d9221f0cd062e8c59fb880be651ebe48cb2182116c9e07fbc1f364794e8b28dcR342-R351)

### Storybook Enhancements

* Added a new Storybook story to demonstrate a minimum password requirement for encrypted files, including a template that handles password validation and error display.
* Clarified labels for max and min file size stories for better user understanding. [[1]](diffhunk://#diff-341837f96fcac157ad5b0e2205276b21eeaa563d9f6a2ea61a904bd44ba20f54L215-R262) [[2]](diffhunk://#diff-341837f96fcac157ad5b0e2205276b21eeaa563d9f6a2ea61a904bd44ba20f54L228-R275)

### Test and Code Quality

* Improved test code by removing unnecessary checks and updating selectors for clarity and robustness. [[1]](diffhunk://#diff-d9221f0cd062e8c59fb880be651ebe48cb2182116c9e07fbc1f364794e8b28dcL187-L188) [[2]](diffhunk://#diff-d9221f0cd062e8c59fb880be651ebe48cb2182116c9e07fbc1f364794e8b28dcR342-R351)

These changes collectively enhance the usability, accessibility, and testability of the `va-file-input` component, while also providing developers with better tools and examples for implementation and validation.

## Related tickets and links

Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/123070

## Testing and review

Testing environments:
- MacOS VoiceOver (Safari, Chrome, and Firefox)
- Windows 11 NVDA (Chrome, Firefox)
I tested the following interactions:

  ### Upload via click/keyboard
    -  Error
       - Empty file
       - File too big
       - File too small
    - Password input initial
    - Password input error
    - File deletion modal
      - Did not delete
  ### Upload via drag & drop
    -  Error
        - Empty file
        - File too big
        - File too small
    - Password input initial
    - Password input error

## Approvals
See the QA Checklists section below for suggested approvals. Use your best judgment if additional reviews are needed. When in doubt, request a review.

**Approval groups**

Add approval groups to the PR as needed:

- Engineering: [platform-design-system-fe](https://github.com/orgs/department-of-veterans-affairs/teams/platform-design-system-fe)
- Accessibility: [platform-design-system-a11y](https://github.com/orgs/department-of-veterans-affairs/teams/platform-design-system-a11y)
- Design: [platform-design-system-designer](https://github.com/orgs/department-of-veterans-affairs/teams/platform-design-system-designers)

## QA checklists

Use the QA checklists below as guides, not rules. Not all checklists will apply to every PR but there could be some overlap.

In all scenarios, changes should be fully tested by the author and verified by the reviewer(s); functionality, responsiveness, etc.

<details>
  <summary>✨ New Component Added</summary>

- [ ] The PR has the `minor` label
- [ ] The component matches the [Figma](https://www.figma.com/files/1499394822283304153/project/105082786?fuid=1192586511403544015) designs.
- [ ] All properties, custom events, and utility functions have e2e and/or unit tests
- [ ] A new Storybook page has been added for the component
- [ ] Tested in all [VA breakpoints](https://design.va.gov/foundation/breakpoints).
- [ ] Chromatic UI Tests have run and snapshot changes have been accepted by the design reviewer
- [ ] Tested in vets-website using [Verdaccio](https://github.com/department-of-veterans-affairs/component-library?tab=readme-ov-file#local-testing-in-vets-website-with-verdaccio)
- [ ] **Engineering** has approved the PR
- [ ] **Design** has approved the PR
- [ ] **Accessibility** has approved the PR
</details>

<details>
  <summary>🌱 New Component Variation Added</summary>

- [ ] The PR has the `minor` label
- [ ] The variation matches its [Figma](https://www.figma.com/files/1499394822283304153/project/105082786?fuid=1192586511403544015) design.
- [ ] Any new properties, custom events, or utility functions have e2e and/or unit tests
- [ ] A new story has been added to the component's existing Storybook page
- [ ] Any Chromatic UI snapshot changes have been accepted by a design reviewer
- [ ] Tested in vets-website using [Verdaccio](https://github.com/department-of-veterans-affairs/component-library?tab=readme-ov-file#local-testing-in-vets-website-with-verdaccio)
- [ ] **Engineering** has approved the PR
- [ ] **Design** has approved the PR
</details>

<details>
  <summary>🐞 Component Fix</summary>

- [ ] The PR has the `patch` label
- [ ] Any new properties, custom events, or utility functions have e2e and/or unit tests
- [ ] Any markup changes are evaluated for impact on vets-website.
    - Will any vets-website tests fail from the change?
- [ ] Any Chromatic UI snapshot changes have been reviewed and approved by a designer if necessary
- [ ] **Engineering** has approved the PR
</details>

<details>
  <summary>♿️ Component Fix - Accessibility</summary>

- [ ] The PR has the `patch` label
- [ ] Any new properties, custom events, or utility functions have e2e and/or unit tests
- [ ] Any Chromatic UI snapshot changes have been reviewed and approved by a designer if necessary
- [ ] **Engineering** has approved the PR
- [ ] **Accessibility** has approved the PR
</details>

<details>
  <summary>🚨 Component Fix - Breaking API Change</summary>

- [ ] The PR has the `major` label
- [ ] vets-website and content-build have been evaluated to determine the impact of the breaking change
- [ ] Any new properties, custom events, or utility functions have e2e and/or unit tests
- [ ] Any Chromatic UI snapshot changes have been reviewed and approved by a designer if necessary
- [ ] Tested in vets-website using [Verdaccio](https://github.com/department-of-veterans-affairs/component-library?tab=readme-ov-file#local-testing-in-vets-website-with-verdaccio)
- [ ] **Engineering** has approved the PR
</details>

<details>
  <summary>🔧 Component Update - Non-Breaking API Change</summary>

- [ ] The PR has the `minor` label
- [ ] Any new properties, custom events, or utility functions have e2e and/or unit tests
- [ ] Any Chromatic UI snapshot changes have been reviewed and approved by a designer if necessary
- [ ] **Engineering** has approved the PR
</details>

<details>
  <summary>📖 Storybook Update</summary>

- [ ] The PR has the `ignore-for-release` label
- [ ] Any Chromatic UI snapshot changes have been reviewed and approved by a designer if necessary
- [ ] **Engineering** has approved the PR
</details>

<details>
  <summary>🎨 CSS-Library Update</summary>

- [ ] The PR has the `css-library` label
- [ ] vets-website and content-build have been checked to determine the impact of any breaking changes
- [ ] **Engineering** has approved the PR
</details>

### CODE REVIEW

**Review by jeana-adhoc — CHANGES_REQUESTED** [review]
Hi @RyanMunsch  - I haven't completed all the testing because I've found some bugs with JAWS & NVDA that I think we need to look into. You identified something simiilar I think in your testing.  [ My results are documented here in detail ](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/platform/design-system/components/va-file-input/tests/accessibility-tests.md)

But it seems that focus is being lost in windows specifically with the drag and drop option. JAWS is announcing....

> dash va dash file dash input dash dash accepts dash only dash specific dash file dash types dash Microsoft edge
> dash va dash file dash input dash dash accepts dash only dash specific dash file dash types dash Microsoft edge page
> dash va dash file dash input dash dash accepts dash only dash specific dash file dash types dash type
> Change file error.
> We do not accept dot mp4 files.
> Choose a new file button to activate press enter. 

https://github.com/user-attachments/assets/c14f08f2-160a-4ec4-add9-33936cceb03b

Even when using the link in Edge + JAWS it always seems to be announcing the `<title>` first. 

@amyleadem I don't think this is what we would want here?

**Comment by RyanMunsch** [comment]
@powellkerry, there have been lots of changes so you last reviewed. One thing to note is that we are considering this a "good enough" improvement, but it is not ideal for each combination of browser and screen reader. There will still be some inconsistencies with Safari & VoiceOver, but in general things should be working better.

File: packages/web-components/src/components/va-file-input/va-file-input.tsx
[the remark this answers is not in the document]
Comment (jeana-adhoc) [reply packages/web-components/src/components/va-file-input/va-file-input.tsx]: > Are we making changes to be consistent across browsers when the user might expect the messaging to be a certain way in the browser they are using? Will these changes make it less accessible because the user is expecting things to be read a different way?

@powellkerry  This was a great callout and thanks for paying attention to this. Thankfully - we're not trying to standardize/change the native reading of things. We won't ever be able to make the screen readers all sound the same. They're reading natively in the same default order users expect

File: packages/web-components/src/components/va-text-input/va-text-input.tsx [hunk 1]
```diff
@@ -533,7 +533,7 @@ export class VaTextInput {
             </label>
           )}
           <slot></slot>
-          <span id="input-error-message" role="alert" class={errorClass}>
+          <span id="input-error-message" role="alert" class={errorClass} tabIndex={-1}>
```
Comment (amyleadem) [inline packages/web-components/src/components/va-text-input/va-text-input.tsx]: **question:** Why add `tabindex=-1` here in all instances?

File: packages/web-components/src/components/va-file-input/utils.tsx [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+ */
+export function focusOnChangeButton(componentHostElement: HTMLElement): void {
+  setTimeout(() => {
+    const changeButton: HTMLElement = componentHostElement.shadowRoot.querySelector('va-button-icon');
+    const innerButton: HTMLElement = changeButton?.shadowRoot.querySelector('button');
+
+    if (innerButton) {
+      innerButton.focus();
+    }
+  }, 250);
```
Comment (amyleadem) [inline packages/web-components/src/components/va-file-input/utils.tsx]: **question:** Just to confirm, could requestAnimationFrame work here instead of an arbitrary timeout?

File: packages/web-components/src/components/va-file-input/utils.tsx [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+/**
+ * Renderer for the displayed error alert.
+ * @returns {HTMLSpanElement | void}
+ */
+export function renderErrorAlert(displayError: string): HTMLSpanElement | void {
+  if (!displayError) return;
+
+  return (
+    <span id="input-error-message" class="usa-error-message">
```
Comment (amyleadem) [inline packages/web-components/src/components/va-file-input/utils.tsx]: **suggestion:** It would be helpful to have a comment about how we are specifically excluding the alert role/aria-live from this error message because we are announcing it via other means. Just wanting to make sure we don't add it in again without consideration.

File: packages/storybook/stories/va-file-input-uswds.stories.tsx [hunk 4]
```diff
@@ -424,7 +472,35 @@ const FileUploadedTemplate = args => {
     };
   }, []);
 
-  return <Template {...args} value={mockFile} />;
+  const handleToggleFileClick = async () => {
```
Comment (amyleadem) [inline packages/storybook/stories/va-file-input-uswds.stories.tsx]: **question:** What is this story testing? Can you add a comment so we remember?

File: packages/web-components/src/components/va-file-input/va-file-input.tsx [hunk 5]
```diff
@@ -510,55 +549,15 @@ export class VaFileInput {
...
-    } else if (fileContents) {
-      if (fileType.startsWith('image/')) {
-        fileThumbnail = (
-          <div class="thumbnail-container" aria-hidden="true">
-            <img class="thumbnail-preview" src={fileContents} alt="image" />
-          </div>
-        );
-      }
-    }
-
```
Comment (amyleadem) [inline packages/web-components/src/components/va-file-input/va-file-input.tsx]: **Reviewer note (no action needed):** Confirmed that this function was moved to utils.tsx with improvements.

File: packages/storybook/stories/va-file-input-uswds.stories.tsx (hunk 4 above)
Comment (amyleadem) [inline packages/storybook/stories/va-file-input-uswds.stories.tsx]: **issue:** It isn't immediately clear why we need these buttons in this story. Can you add a comment explaining the purpose?

File: packages/web-components/src/components/va-file-input/utils.tsx [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+  // First requestAnimationFrame: Queues for next frame (after Stencil render)
+  // Second requestAnimationFrame: Ensures layout and paint have completed
+  requestAnimationFrame(() => {
+    requestAnimationFrame(() => {
+      if (fileInputEl) {
+        fileInputEl.focus();
+      }
+    });
+  });
+}
```
Comment (amyleadem) [inline packages/web-components/src/components/va-file-input/utils.tsx]: **issue:** I noticed that focus is moving to the component on page load in the [docs](https://file-input-focus-and-sr--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/docs/uswds-va-file-input--docs) and [uploaded file stories](https://file-input-focus-and-sr--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/story/uswds-va-file-input--upload-status). This felt unexpected to me, where I expected the component to only receive focus after user action (either closing the file selection dialog or interacting with the component). Was this intentional?

File: packages/web-components/src/components/va-file-input/utils.tsx [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+  }
+  else if (initialUploadAttemptHasTakenPlace && !file && !uploadedFile) {
+    inputAriaLabel = `File deleted. No file selected. ${inputAriaLabel}`;
+  }
+
+  let changeFileAriaLabel: string | undefined = undefined;
+  let deleteFileAriaLabel: string | undefined = undefined;
+
+  if (file) {
+    changeFileAriaLabel = `change file ${file.name}`;
```
Comment (amyleadem) [inline packages/web-components/src/components/va-file-input/utils.tsx]: **nitpick:** Let's add colons here - some SRs will announce it and it gives just a tad bit more clarity
```suggestion
    changeFileAriaLabel = `Change file: ${file.name}`;
```

File: packages/web-components/src/components/va-file-input/utils.tsx [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+  if (displayError) {
+    return {
+      inputAriaLabel: `Error: ${displayError}. ${inputAriaLabel}`,
+      deleteFileAriaLabel: `delete file. Error. ${displayError}`,
+      changeFileAriaLabel: `change file. Error. ${displayError}`,
+    }
+  }
+
+  if (initialUploadAttemptHasTakenPlace && file) {
+    inputAriaLabel = `You have selected the file: ${file.name}.`;
```
Comment (amyleadem) [inline packages/web-components/src/components/va-file-input/utils.tsx]: **note:** In Chrome/VO, the file name is announced twice here because it seems like the browser attaches the name of the selected file to the input by default. This duplicate readout is ok in my mind because Safari/VO would not announce the file name without it.

**Review by amyleadem — CHANGES_REQUESTED** [review]
Preliminary tests are looking really good! Thanks a ton for your work in organizing your code and writing very through and helpful comments. It helped tell the story of the changes really well. 

I had [one major issue/question related to focus management in the component on page load](https://github.com/department-of-veterans-affairs/component-library/pull/1886#discussion_r2670365846) (details in the comments), along with some minor nitpicks/questions in the comments below. 

So far I have only done preliminary interaction testing in VO/Chrome and VO/Safari but I'll pick up more tomorrow. Let me know if you have questions!

File: packages/web-components/src/components/va-text-input/va-text-input.tsx (hunk 1 above)
Comment (RyanMunsch) [reply packages/web-components/src/components/va-text-input/va-text-input.tsx]: Nice catch. It turns out that we don't need this at all. I'm pretty sure that this is a remnant from earlier iterations when I was struggling to get announcements working without manually setting focus. I tested on VoiceOver (MacOS and iOS) and NVDA and focus is still working as expected after removing the `tabIndex`.

File: packages/web-components/src/components/va-file-input/utils.tsx (hunk 2 above)
Comment (RyanMunsch) [reply packages/web-components/src/components/va-file-input/utils.tsx]: As discussed on slack, I created a generic helper function to set focus with retries and call it for both the change button and password input.

File: packages/web-components/src/components/va-file-input/va-file-input.tsx [hunk 9]
```diff
@@ -570,142 +569,149 @@ export class VaFileInput {
...
+        */}
         <div class="file-input-wrapper" onDrop={this.handleDrop}>
           <input
             id="fileInputField"
             class="file-input"
-            aria-label={`${label}${required ? ' ' + i18next.t('required') : ''}. ${dragFileString}${chooseFileString}`}
-            style={{
-              visibility: (uploadStatus === 'success' || uploadedFile || internalError) ? 'hidden' : 'unset',
-            }}
+            aria-label={inputAriaLabel}
```
[the remark this answers is not in the document]
Comment (RyanMunsch) [reply packages/web-components/src/components/va-file-input/va-file-input.tsx]: @amyleadem, in the init state, the label is included in the input's aria-label, as its passed to the helper method and included in the "default" value.

- [Passed to util function](https://github.com/department-of-veterans-affairs/component-library/blob/752846394b5d685d11cc74813b170a705cb444d6/packages/web-components/src/components/va-file-input/va-file-input.tsx#L578)
- [Included in default value](https://github.com/department-of-veterans-affairs/component-library/blob/752846394b5d685d11cc74813b170a705cb444d6/packages/web-components/src/components/va-file-input/utils.tsx#L146)

Monosnap Image 2026-01-08 14-02-25

From what I'm seeing, Safari behaves differently from Chrome in that it announces the native shadow content of the `<input>` element as priority before eventually announcing the aria-label, including the passed label:

https://github.com/user-attachments/assets/45833366-b977-4332-81ac-9bc3adf9e3da

I'm not sure if there's anything we can do about this without creating additional problems. In theory, I think we could have `visibility: hidden` on the <input> element temporarily, but as we have seen that creates issues in general and may lead to edge cases where focus does not work as expected at all for Safari.

Let me know what you think!

[the remark this answers is not in the document]
Comment (amyleadem) [reply packages/web-components/src/components/va-file-input/va-file-input.tsx]: Thanks for the explanation. I think at least part of this was me having some review fatigue and I honestly thought that the text wasn't included in the aria-label, but looking at it fresh, I see it there. It looks like this is a pre-existing issue in Safari, so let's just break this into a follow-on issue.

**Review by it-harrison — APPROVED** [review]
@RyanMunsch looks good! I tested this in vets-website and encountered no problems and all e2e tests passed.
